### PR TITLE
☀️ Common: 유효성 체크를 직접 관리하기 위한 Preconditions 유틸리티 추가

### LINKED ISSUE #138: ☀️ Common: 유효성 체크를 직접 관리하기 위한 `Preconditions` 유틸리티 추가
### 이슈 내용

Each validator will offer overloads with optional `Throwable` causes, `Runnable` callbacks, and `Supplier<Map<String, Object>>` payloadSuppliers.

### <br /> 부연 설명

우리가 사용하는 사용자 입력 유효성 체크 기능은 spring validation 및 hibernate 등에서 충분히 지원합니다.
하지만 커스텀에 몇 가지 제약을 갖고 있으므로, 자유로운 제어를 위해 가끔 유효성을 직접 핸들링하게 됩니다.

`common` 모듈에 이를 위한 유틸리티 클래스를 생성하여 관리합니다.

### LINKED ISSUE #139: [SUB-ISSUE] Common > Preconditions: Not Null 체크 유틸리티 함수 목록 추가 (Object)

### LINKED ISSUE #140: [SUB-ISSUE] Common > Preconditions: Not Empty 체크 유틸리티 함수 목록 추가 (String)

### LINKED ISSUE #141: [SUB-ISSUE] Common > Preconditions: Not Blank 체크 유틸리티 함수 목록 추가 (String)

### LINKED ISSUE #142: [SUB-ISSUE] Common > Preconditions: String Length Range 체크 유틸리티 함수 목록 추가

### LINKED ISSUE #143: [SUB-ISSUE] Common > Preconditions: String Min Length 체크 유틸리티 함수 목록 추가

### LINKED ISSUE #144: [SUB-ISSUE] Common > Preconditions: String Max Length 체크 유틸리티 함수 목록 추가

### LINKED ISSUE #145: [SUB-ISSUE] Common > Preconditions: Collection Length Range 체크 유틸리티 함수 목록 추가

### LINKED ISSUE #146: [SUB-ISSUE] Common > Preconditions: Collection Min Length 체크 유틸리티 함수 목록 추가

### LINKED ISSUE #147: [SUB-ISSUE] Common > Preconditions: Collection Max Length 체크 유틸리티 함수 목록 추가

### LINKED ISSUE #148: [SUB-ISSUE] Common > Preconditions: String 정규표현식 체크 유틸리티 함수 목록 추가

### PR DESCRIPTION
**Pull request by merge-simpson**
## Issues

- Resolves #138 
  - Resolves #139 
  - Resolves #140 
  - Resolves #141 
  - Resolves #142 
  - Resolves #143 
  - Resolves #144 
  - Resolves #145 
  - Resolves #146 
  - Resolves #147 
  - Resolves #148 

## Description

- 유효성을 커스텀하는 등 직접 핸들링해야 할 때를 위한 유틸리티 클래스를 추가합니다.
  - `spring-boot-starter-validation`, hibernate 유효성 등과 사용감에서 나타나는 차이는 대표적으로 다음과 같습니다.
    - `strip`이나 `substring`, `replaceAll` 등 문자열을 전처리한 후 길이, 정규식 등 유효성을 체크할 때, 이처럼 커스텀된 메서드와 조합하면 편리합니다.
    - 반환하는 예외를 `CustomException` 등 우리가 원하는 런타임 예외 종류로 지정할 수 있습니다.

<br />

## Review Points

- 한 번에 작업했다 보니 코드가 길어 리뷰가 까다롭습니다.
- 위 연결된 이슈 목록을 살피고, 부족한 유효성 항목이 있다면 새로운 이슈로 등록해 주시면 감사드리겠습니다.  
    : 예를 들어 시간 관련 유효성도 추가할 수 있지만, 사용처가 제한되므로 아직 작성하지 않는 것이 낫다고 보았습니다.

<br />

## How Has This Been Tested?

- 분량이 많지만, 오류를 발견한 후에도 수정할 수 있으므로 이번에는 테스트 생략했습니다.

<br />

## Additional Notes

**간단한 용례**

static import는 일반적으로 삼가는 관례를 따르지만, 이처럼 재사용성이 높은 일부 유틸리티 함수는 static import를 허용하곤 합니다.

```java
@Builder
public record BlogCreateCommand(
        String userId,
        String name,
        String url
) {
    public BlogCreateCommand {
        validateNotBlank(userId, BLOG_OWNER_ID_REQUIRED);
        validateNotBlank(name, BLOG_NAME_REQUIRED);
        validateNotBlank(url, BLOG_URL_REQUIRED);

        // 앞뒤 공백 문자를 제거 후 유효성 확인
        name = name.strip();
        url = url.strip();

        validateLength(name, 3, 30, BLOG_NAME_INVALID_LENGTH);
        validateRegex(url, "^[A-Za-z0-9_-]+$", BLOG_URL_INVALID_FORMAT);
        validateLength(url, 3, 15, BLOG_URL_INVALID_LENGTH);
    }
}
```

### CODE REVIEW

**Comment by merge-simpson** [comment]
@wch-os 님네 코파일럿 성능이 탐나서 코파일럿 리뷰를 추가하고 싶었는데 저는 이제 버튼이 안 보이네요.

대신 chat with copilot으로 리뷰가 가능한지 물었습니다! 다행히 PR 통째로 스캔하는 기능은 쓸 수 있네요.

<br />

---

<a id="copilot-review-text-english"> </a>

English | [한국어](#copilot-review-text-korean)

<br />

Based on the available information, this PR introduces a new utility class called Preconditions for custom validation handling. Let me break down the key points and provide a review:

## Key Points:

The PR adds a new utility class Preconditions.java in the common/src/main/java/nettee/common/validation/ directory
It resolves multiple related issues (#138-#148)
The purpose is to provide custom validation handling when direct control over validation is needed
The changes include 865 additions with no deletions, indicating this is a new feature addition
Advantages of the proposed solution:

Allows custom string preprocessing before validation (strip, substring, replaceAll)
Enables throwing custom runtime exceptions instead of standard validation exceptions
Provides a unified utility for validation across the codebase

## Review Comments:

### Scope and Size:

The PR is quite large with 865 lines of additions
It would have been better to split this into smaller, more focused PRs for easier review
However, since all changes are related to validation utilities, it's still manageable

### Testing:

The author mentions that tests were skipped for this PR
RECOMMENDATION: Given the utility nature of this class and its potential widespread usage, unit tests should be added to ensure reliability
Suggest adding tests covering different validation scenarios and edge cases
Documentation:

The PR description clearly explains the motivation and benefits
Would be helpful to see some usage examples in the code documentation
RECOMMENDATION: Add JavaDoc comments with examples for the main validation methods

### Design:

The approach of having a dedicated validation utility aligns with good practice
The flexibility to use custom exceptions is a good design choice
String preprocessing capabilities add valuable functionality
Future Considerations:

The author notes that time-related validations were intentionally omitted due to limited use cases
This is a reasonable decision, following the YAGNI principle
Additional validations can be added later as needed

## Suggestions:

Add comprehensive unit tests before merging
Consider adding code examples in the PR description or documentation
Include JavaDoc documentation for the public methods
Consider adding a README or documentation file explaining common usage patterns
Overall, this appears to be a valuable addition to the codebase, providing needed flexibility in validation handling. The main concern is the lack of tests, which should be addressed before merging.

Would you like me to focus on any specific aspect of the review or provide more detailed recommendations in any particular area?

<br />

---

<a id="copilot-review-text-korean"> </a>

[English](#copilot-review-text-english) | 한국어

<br />

사용 가능한 정보를 바탕으로, 이 PR은 사용자 정의 검증 처리를 위한 새로운 유틸리티 클래스인 Preconditions를 도입합니다. 주요 포인트를 정리하고 검토를 제공하겠습니다:

## 주요 포인트:

이 PR은 common/src/main/java/nettee/common/validation/ 디렉토리에 새로운 유틸리티 클래스 Preconditions.java를 추가합니다
이 클래스는 관련 문제(#138-#148)를 해결합니다
목적은 직접적인 검증 제어권이 필요한 경우 맞춤형 검증 처리를 제공하는 것입니다.  
변경 사항은 865개의 추가 행으로 구성되며 삭제 행은 없으며, 이는 새로운 기능 추가를 의미합니다.  
제안된 솔루션의 장점:  

검증 전 맞춤형 문자열 사전 처리(strip, substring, replaceAll)를 허용합니다.  
표준 검증 예외 대신 맞춤형 런타임 예외를 던질 수 있습니다.  
코드베이스 전체에서 검증에 대한 통합된 유틸리티를 제공합니다.  

## 검토 의견

### 범위 및 규모

이 PR은 865줄의 추가로 상당히 대규모입니다  
더 작은 규모로 분할하여 검토를 용이하게 하는 것이 더 나았을 것입니다  
그러나 모든 변경 사항이 검증 유틸리티와 관련되어 있으므로 여전히 관리 가능합니다  

### 테스트:

저자는 이 PR에 대한 테스트가 생략되었다고 언급했습니다
권장 사항: 이 클래스의 유틸리티 특성 및 광범위한 사용 가능성을 고려할 때 신뢰성을 확보하기 위해 유닛 테스트를 추가해야 합니다
다양한 검증 시나리오와 경계 사례를 커버하는 테스트를 추가하는 것을 권장합니다

### 문서화:

PR 설명은 동기와 이점을 명확히 설명합니다
코드 문서화에 사용 예시를 포함하면 도움이 될 것입니다
권장 사항: 주요 검증 메서드에 예시를 포함한 JavaDoc 주석을 추가하세요

### 설계:

전용 검증 유틸리티를 사용하는 접근 방식은 좋은 관행과 일치합니다
맞춤형 예외를 사용할 수 있는 유연성은 좋은 설계 선택입니다
문자열 사전 처리 기능은 유용한 기능을 추가합니다

### 미래 고려 사항:

저자는 시간 관련 검증들이 제한된 사용 사례로 인해 의도적으로 생략되었다고 언급합니다
이는 YAGNI 원칙에 따라 합리적인 결정입니다
필요에 따라 추가 검증을 나중에 추가할 수 있습니다

## 제안 사항:

머지 전에 포괄적인 유닛 테스트를 추가하세요
PR 설명이나 문서에 코드 예제를 추가하는 것을 고려하세요
공개 메서드에 대한 JavaDoc 문서를 포함하세요
일반적인 사용 패턴을 설명하는 README 또는 문서 파일을 추가하는 것을 고려하세요
전체적으로 이 코드베이스에 필요한 유연성을 제공하는 가치 있는 추가 사항으로 보입니다. 주요 우려 사항은 테스트 부족으로, 병합 전에 해결되어야 합니다.

리뷰의 특정 측면에 집중하거나 특정 분야에서 더 자세한 권장 사항을 제공해 드릴까요?

_Translated with [DeepL.com](https://www.deepl.com/ko/translator) (free version)_

File: common/src/main/java/nettee/common/validation/Preconditions.java [hunk 1]
```diff
@@ -0,0 +1,865 @@
...
+public final class Preconditions {
+    private Preconditions() {}
+
+    // ╭────────────────────────────╮
+    //    validateNotNull variants
+    // ╰────────────────────────────╯
+
+    public static void validateNotNull(Object value, ErrorCode errorCode) {
+        if (value == null) {
+            throw errorCode.exception();
```
Comment (sweetykr7) [inline common/src/main/java/nettee/common/validation/Preconditions.java:20]: 단순 null 체크를 이렇게 쓰는 거군요. throw errorCode.exception()를 통해 예외를 간결하게 처리하는 방식이 눈에 띄었습니다

File: common/src/main/java/nettee/common/validation/Preconditions.java [hunk 2]
```diff
@@ -0,0 +1,865 @@
...
+            ErrorCode errorCode,
+            Runnable runnable,
+            Throwable cause
+    ) {
+        if (isEmpty(value)) {
+            throw errorCode.exception(runnable, cause);
+        }
+    }
+
+    public static void validateNotEmpty(
```
Comment (sweetykr7) [inline common/src/main/java/nettee/common/validation/Preconditions.java:155]: 예외에 payload를 공급하는 방식으로 이렇게 쓰는 거군요. Supplier를 통해 예외에 추가 데이터를 유연하게 전달하는 설계가 돋보였습니다.

File: common/src/main/java/nettee/common/validation/Preconditions.java [hunk 3]
```diff
@@ -0,0 +1,865 @@
...
+        }
+    }
+
+    public static void validateNotBlank(String value, ErrorCode errorCode, Throwable cause) {
+        if (isBlank(value)) {
+            throw errorCode.exception(cause);
+        }
+    }
+
+    public static void validateNotBlank(String value, ErrorCode errorCode, Runnable runnable) {
```
Comment (sweetykr7) [inline common/src/main/java/nettee/common/validation/Preconditions.java:192]: 런타임 시점의 실행 로직을 예외 생성과 함께 전달할 수 있게 이렇게 쓰는 거군요. Runnable을 통해 예외 처리 시 부가 동작을 실행하는 아이디어가 인상 깊었습니다.

File: common/src/main/java/nettee/common/validation/Preconditions.java [hunk 4]
```diff
@@ -0,0 +1,865 @@
...
+            return; // 빈 문자열도 min == 0이면 허용
+        }
+
+        int len = value.length();
+        if (len < min || len > max) {
+            throw exceptionSupplier.get();
+        }
+    }
+
+    private static void performLengthValidation(
```
Comment (sweetykr7) [inline common/src/main/java/nettee/common/validation/Preconditions.java:696]: 길이 검증 로직을 공통 메서드로 이렇게 쓰는 거군요. Supplier를 통해 예외 생성을 지연시켜 중복 코드를 줄이는 구성이 좋았습니다.

**Review by wch-os — APPROVED** [review]
유효성 체크 코드가 어떻게 보면 정말 짧은 코드이지만 계속 반복해서 사용되는 코드인데, 이렇게 유틸리티 클래스로 관리될 수 있겠군요!

지금까지 각 라이브러리에서 지원해주는 유틸 클래스만 사용할 줄 알았지, 이렇게 프로젝트 유틸로 관리할 생각은 못했는데 새로운 방법에 대해 제안해주셔 감사합니다!

유연성 있는 설계를 위해 고민이 많으셨을 텐데, 수고 많으셨습니다!

---

> 문서에 코드 예제를 추가하는 것을 고려하세요
공개 메서드에 대한 JavaDoc 문서를 포함하세요
일반적인 사용 패턴을 설명하는 README 또는 문서 파일을 추가하는 것을 고려하세요

💊 추가로 copilot의 제안대로 공개 메서드로 사용되는 기능이므로, 일부 문서나 주석이 조금 더 들어가면 좋을 것 같습니다!

File: common/src/main/java/nettee/common/validation/Preconditions.java [hunk 5]
```diff
@@ -0,0 +1,865 @@
...
+        if (collection.isEmpty()) {
+            return; // 항상 <= max (where max >= 0)
+        }
+
+        if (collection.size() > max) {
+            throw exceptionSupplier.get();
+        }
+    }
+
+    private static void performRegexValidation(
```
Comment (sweetykr7) [inline common/src/main/java/nettee/common/validation/Preconditions.java:806]: 정규식을 검증하는 로직을 이렇게 쓰는 거군요. Pattern.compile과 matcher.matches()를 사용해 유효성을 명확히 확인하는 구현이 눈에 띄었습니다.

File: common/src/main/java/nettee/common/validation/Preconditions.java [hunk 6]
```diff
@@ -0,0 +1,865 @@
...
+        assert min >= 0 : "min cannot be less than 0";
+        assert max >= 0 : "max cannot be less than 0";
+        assert min <= max : "max must be greater than or equal to " + min;
+
+        if (value == null && min != 0) {
+            throw new NullPointerException("String must not be null if min is not 0.");
+        }
+    }
+
+    private static void validateLengthArgument(Collection<?> collection, int min, int max) {
```
Comment (sweetykr7) [inline common/src/main/java/nettee/common/validation/Preconditions.java:829]: 길이 검증 전에 파라미터를 엄격하게 검사하도록 이렇게 쓰는 거군요. assert를 사용해 잘못된 인자를 조기에 잡는 방식이 좋았습니다~!
